compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -1522,7 +1522,8 @@ private boolean checkSimilarMatchGuard(BLangMatchGuard firstMatchGuard, BLangMat
             BLangTypeTestExpr secondTypeTest = (BLangTypeTestExpr) secondMatchGuard.expr;
             return ((BLangSimpleVarRef) firstTypeTest.expr).variableName.toString().equals(
                     ((BLangSimpleVarRef) secondTypeTest.expr).variableName.toString()) &&
-                    firstTypeTest.typeNode.getBType().tag == secondTypeTest.typeNode.getBType().tag;
+                    types.isAssignable(firstTypeTest.typeNode.getBType(),
+                            secondTypeTest.typeNode.getBType());
```

Comment:
Actually, checking just assignability is not enough. We need to consider effective types/narrowed types. We can do this when fixing match type narrowing.

```ballerina
public function main() {
    int|boolean a = 1;

    match a {
        var x if a is int|string => {

        }
        var y if a is int|int[] => {

        }
    }
}
```

The second match clause here is unreachable.

Shall we create an issue for this and point to https://github.com/ballerina-platform/ballerina-lang/issues/30191?

This particular example results in an error btw, created a separate issue for this - https://github.com/ballerina-platform/ballerina-lang/issues/31587.

Reply:
Created https://github.com/ballerina-platform/ballerina-lang/issues/31588.